File: TPs/tp3/extra/extras.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <openssl/md5.h>

// Constantes para el filtro de conteo
#define INITIAL_SIZE 1000
#define NUM_HASHES 3
#define RESIZE_THRESHOLD 0.1
#define RESIZE_FACTOR 2

typedef struct {
    int size;
    int num_hashes;
    double resize_threshold;
    int resize_factor;
    int *filter;
    unsigned int *hash_seeds;
    int element_count;
} countfilter_t;

typedef struct {
    void **elements;
    int count;
    int capacity;
    countfilter_t *filter;
    size_t data_type_size;
    unsigned char *(*hash_function)(const unsigned char *, size_t, unsigned char *);
    int (*compare_function)(const void *, const void *);
} unique_list_t;

unsigned int *generate_seeds(int num_hashes) {
    unsigned int *seeds = malloc(num_hashes * sizeof(unsigned int));
    for (int i = 0; i < num_hashes; i++) {
        seeds[i] = rand();
    }
    return seeds;
}

unsigned char *md5_hash(const unsigned char *data, size_t length, unsigned char *result) {
    MD5(data, length, result);
    return result;
}

int *get_indices(countfilter_t *filter, unsigned char *hashes) {
    int *indices = malloc(filter->num_hashes * sizeof(int));
    for (int i = 0; i < filter->num_hashes; i++) {
        indices[i] = (hashes[i * MD5_DIGEST_LENGTH] + filter->hash_seeds[i]) % filter->size;
    }
    return indices;
}

countfilter_t *create_count_filter(int size, int num_hashes, double resize_threshold, int resize_factor) {
    countfilter_t *filter = malloc(sizeof(countfilter_t));
    if (!filter) return NULL;
    filter->size = size;
    filter->num_hashes = num_hashes;
    filter->resize_threshold = resize_threshold;
    filter->resize_factor = resize_factor;
    filter->filter = calloc(size, sizeof(int));
    if (!filter->filter) {
        free(filter);
        return NULL;
    }
    filter->hash_seeds = generate_seeds(num_hashes);
    filter->element_count = 0;
    return filter;
}

double false_positive_probability(countfilter_t *filter) {
    return pow(1 - exp(-filter->num_hashes * (double)filter->element_count / filter->size), filter->num_hashes);
}

void resize_count_filter(countfilter_t *filter, int new_size) {
    int *old_filter = filter->filter;
    int old_size = filter->size;
    filter->size = new_size;
    filter->filter = calloc(new_size, sizeof(int));
    if (!filter->filter) return;
    filter->hash_seeds = generate_seeds(filter->num_hashes);

    for (int i = 0; i < old_size; i++) {
        if (old_filter[i] > 0) {
            char element[12];
            snprintf(element, 12, "%d", i);
            unsigned char hash[MD5_DIGEST_LENGTH];
            md5_hash((unsigned char *)element, strlen(element), hash);
            int *indices = get_indices(filter, hash);
            for (int j = 0; j < filter->num_hashes; j++) {
                filter->filter[indices[j]] += old_filter[i];
            }
            free(indices);
        }
    }
    free(old_filter);
}

void add_to_filter(countfilter_t *filter, const char *element) {
    filter->element_count++;
    unsigned char hash[MD5_DIGEST_LENGTH];
    md5_hash((unsigned char *)element, strlen(element), hash);
    int *indices = get_indices(filter, hash);
    for (int i = 0; i < filter->num_hashes; i++) {
        filter->filter[indices[i]]++;
    }
    free(indices);

    if (false_positive_probability(filter) > filter->resize_threshold) {
        resize_count_filter(filter, filter->size * filter->resize_factor);
    }
}

void remove_from_filter(countfilter_t *filter, const char *element) {
    unsigned char hash[MD5_DIGEST_LENGTH];
    md5_hash((unsigned char *)element, strlen(element), hash);
    int *indices = get_indices(filter, hash);
    for (int i = 0; i < filter->num_hashes; i++) {
        if (filter->filter[indices[i]] > 0) {
            filter->filter[indices[i]]--;
        }
    }
    filter->element_count--;
    free(indices);
}

int contains_in_filter(countfilter_t *filter, const char *element) {
    unsigned char hash[MD5_DIGEST_LENGTH];
    md5_hash((unsigned char *)element, strlen(element), hash);
    int *indices = get_indices(filter, hash);
    for (int i = 0; i < filter->num_hashes; i++) {
        if (filter->filter[indices[i]] == 0) {
            free(indices);
            return 0;
        }
    }
    free(indices);
    return 1;
}

unique_list_t *create_unique_list(size_t data_type_size, unsigned char *(*hash_function)(const unsigned char *, size_t, unsigned char *), int (*compare_function)(const void *, const void *)) {
    unique_list_t *list = malloc(sizeof(unique_list_t));
    if (!list) return NULL;
    list->elements = malloc(INITIAL_SIZE * sizeof(void *));
    if (!list->elements) {
        free(list);
        return NULL;
    }
    list->count = 0;
    list->capacity = INITIAL_SIZE;
    list->filter = create_count_filter(INITIAL_SIZE, NUM_HASHES, RESIZE_THRESHOLD, RESIZE_FACTOR);
    list->data_type_size = data_type_size;
    list->hash_function = hash_function;
    list->compare_function = compare_function;
    return list;
}

int insert_to_unique_list(unique_list_t *list, void *element) {
    unsigned char hash[MD5_DIGEST_LENGTH];
    list->hash_function((unsigned char *)element, list->data_type_size, hash);

    if (!contains_in_filter(list->filter, (char *)hash)) {
        if (list->count == list->capacity) {
            list->capacity *= RESIZE_FACTOR;
            list->elements = realloc(list->elements, list->capacity * sizeof(void *));
            if (!list->elements) return 0;
        }
        list->elements[list->count] = malloc(list->data_type_size);
        if (!list->elements[list->count]) return 0;
        memcpy(list->elements[list->count], element, list->data_type_size);
        add_to_filter(list->filter, (char *)hash);
        list->count++;
        return 1;
    } else {
        for (int i = 0; i < list->count; i++) {
            if (list->compare_function(list->elements[i], element)) {
                return 0;
            }
        }
        if (list->count == list->capacity) {
            list->capacity *= RESIZE_FACTOR;
            list->elements = realloc(list->elements, list->capacity * sizeof(void *));
            if (!list->elements) return 0;
        }
        list->elements[list->count] = malloc(list->data_type_size);
        if (!list->elements[list->count]) return 0;
        memcpy(list->elements[list->count], element, list->data_type_size);
        add_to_filter(list->filter, (char *)hash);
        list->count++;
        return 1;
    }
}

int delete_from_unique_list(unique_list_t *list, void *element) {
    unsigned char hash[MD5_DIGEST_LENGTH];
    list->hash_function((unsigned char *)element, list->data_type_size, hash);

    if (contains_in_filter(list->filter, (char *)hash)) {
        for (int i = 0; i < list->count; i++) {
            if (list->compare_function(list->elements[i], element)) {
                free(list->elements[i]);
                for (int j = i; j < list->count - 1; j++) {
                    list->elements[j] = list->elements[j + 1];
                }
                list->count--;
                remove_from_filter(list->filter, (char *)hash);
                return 1;
            }
        }
    }
    return 0;
}

void resize_unique_list_filter(unique_list_t *list, int new_size) {
    resize_count_filter(list->filter, new_size);
}

// Ejemplo de funciones de hash y comparación
unsigned char *example_hash_function(const unsigned char *data, size_t length, unsigned char *result) {
    return md5_hash(data, length, result);
}

int example_compare_function(const void *el1, const void *el2) {
    return strcmp((const char *)el1, (const char *)el2) == 0;
}

// Ejemplo de uso
int main() {
    unique_list_t *unique_list = create_unique_list(sizeof(char *), example_hash_function, example_compare_function);

    char *element1 = "elemento1";
    char *element2 = "elemento2";
    char *element3 = "elemento3";

    insert_to_unique_list(unique_list, element1);
    insert_to_unique_list(unique_list, element2);
    insert_to_unique_list(unique_list, element3);

    delete_from_unique_list(unique_list, element2);

    resize_unique_list_filter(unique_list, 2000);

    for (int i = 0; i < unique_list->count; i++) {
        free(unique_list->elements[i]);
    }
    free(unique_list->elements);
    free(unique_list->filter->filter);
    free(unique_list->filter->hash_seeds);
    free(unique_list->filter);
    free(unique_list);

    return 0;
}

File: TPs/tp3/extra/extras.py
import hashlib
import random
import math

# Constantes para el filtro de conteo
INITIAL_SIZE = 1000
NUM_HASHES = 3
RESIZE_THRESHOLD = 0.1
RESIZE_FACTOR = 2

class CountFilter:
    def __init__(self, size=INITIAL_SIZE, num_hashes=NUM_HASHES, resize_threshold=RESIZE_THRESHOLD, resize_factor=RESIZE_FACTOR):
        self.size = size
        self.num_hashes = num_hashes
        self.resize_threshold = resize_threshold
        self.resize_factor = resize_factor
        self.filter = [0] * size
        self.hash_seeds = [random.randint(0, 2**32 - 1) for _ in range(num_hashes)]
        self.element_count = 0

    def _hashes(self, element):
        return [hashlib.md5((str(seed) + element).encode()).hexdigest() for seed in self.hash_seeds]

    def _indices(self, hashes):
        return [int(h, 16) % self.size for h in hashes]

    def add(self, element):
        self.element_count += 1
        hashes = self._hashes(element)
        indices = self._indices(hashes)
        for index in indices:
            self.filter[index] += 1
        if self._false_positive_probability() > self.resize_threshold:
            self.resize(self.size * self.resize_factor)

    def remove(self, element):
        hashes = self._hashes(element)
        indices = self._indices(hashes)
        for index in indices:
            if self.filter[index] > 0:
                self.filter[index] -= 1
        self.element_count -= 1

    def contains(self, element):
        hashes = self._hashes(element)
        indices = self._indices(hashes)
        return all(self.filter[index] > 0 for index in indices)

    def _false_positive_probability(self):
        return (1 - math.exp(-self.num_hashes * self.element_count / self.size)) ** self.num_hashes

    def resize(self, new_size):
        old_elements = [i for i, count in enumerate(self.filter) if count > 0]
        self.size = new_size
        self.filter = [0] * new_size
        self.hash_seeds = [random.randint(0, 2**32 - 1) for _ in range(self.num_hashes)]
        for element in old_elements:
            self.add(str(element))  # Rehashing elements with new size


class UniqueList:
    def __init__(self, data_type, hash_function, compare_function, filter_size=INITIAL_SIZE, filter_num_hashes=NUM_HASHES, filter_resize_threshold=RESIZE_THRESHOLD, filter_resize_factor=RESIZE_FACTOR):
        self.data_type = data_type
        self.hash_function = hash_function
        self.compare_function = compare_function
        self.elements = []
        self.filter = CountFilter(size=filter_size, num_hashes=filter_num_hashes, resize_threshold=filter_resize_threshold, resize_factor=filter_resize_factor)

    def insert(self, element):
        element_bytes = self.data_type(element)
        element_hash = self.hash_function(element_bytes, len(element_bytes))
        
        if not self.filter.contains(element_hash):
            self.elements.append(element)
            self.filter.add(element_hash)
            return True
        else:
            for el in self.elements:
                if self.compare_function(el, element):
                    return False
            self.elements.append(element)
            self.filter.add(element_hash)
            return True

    def delete(self, element):
        element_bytes = self.data_type(element)
        element_hash = self.hash_function(element_bytes, len(element_bytes))
        
        if self.filter.contains(element_hash):
            for el in self.elements:
                if self.compare_function(el, element):
                    self.elements.remove(el)
                    self.filter.remove(element_hash)
                    return True
        return False

    def resize_filter(self, new_size):
        self.filter.resize(new_size)


# Ejemplo de uso:

def example_hash_function(data, length):
    return hashlib.sha256(data).hexdigest()

def example_compare_function(el1, el2):
    return el1 == el2

# Crear una lista única para almacenar strings
unique_list = UniqueList(lambda x: x.encode('utf-8'), example_hash_function, example_compare_function)

# Insertar elementos
print(unique_list.insert("hello"))  # True
print(unique_list.insert("world"))  # True
print(unique_list.insert("hello"))  # False

# Eliminar elementos
print(unique_list.delete("hello"))  # True
print(unique_list.delete("hello"))  # False

# Redimensionar el filtro de conteo manualmente
unique_list.resize_filter(2000)
